packages/gatsby/src/state-machines/data-layer/index.ts

```diff
@@ -0,0 +1,95 @@
+import { MachineConfig, Machine } from "xstate"
+import { dataLayerActions } from "./actions"
+import { IDataLayerContext } from "./types"
+import { dataLayerServices } from "./services"
+
+export type DataLayerResult = Pick<
+  IDataLayerContext,
+  | "gatsbyNodeGraphQLFunction"
+  | "graphqlRunner"
+  | "pagesToBuild"
+  | "pagesToDelete"
+>
+
+export const dataLayerStates: MachineConfig<IDataLayerContext, any, any> = {
+  initial: `customizingSchema`,
+  states: {
+    customizingSchema: {
+      invoke: {
+        src: `customizeSchema`,
+        id: `customizing-schema`,
+        onDone: {
+          target: `sourcingNodes`,
+        },
+      },
+    },
+    sourcingNodes: {
+      invoke: {
+        src: `sourceNodes`,
+        id: `sourcing-nodes`,
+        onDone: {
+          target: `buildingSchema`,
+          actions: `assignChangedPages`,
+        },
+      },
+    },
+    buildingSchema: {
+      invoke: {
+        id: `building-schema`,
+        src: `buildSchema`,
+        onDone: {
+          target: `creatingPages`,
+          actions: `assignGatsbyNodeGraphQL`,
+        },
+      },
+    },
+    creatingPages: {
+      invoke: {
+        id: `creating-pages`,
+        src: `createPages`,
+        onDone: [
+          {
+            target: `creatingPagesStatefully`,
+            actions: `assignChangedPages`,
+            cond: (context): boolean => !!context.firstRun,
+          },
+          {
+            target: `done`,
+            actions: `assignChangedPages`,
```

Comment:
Wouldn't this (what is this object called?) have the opposite condition as its sibling?

Couldn't `creatingPagesStatefully` detect this condition and be a noop if that's the case? 
- Not sure about this myself, actually. A question of where the logic ought to be.
  - Actually inclined to say that it's not bad to have this in the state config, but
  - At the same time one might argue it should have as little logic as possible. However, 
  - If all these states represents Functions at some point, spinning up a Function just for the sake of it being a noop might be unnecessarily expensive.

Thoughts?

Reply:
Logic that affects flow should ideally be in the machine, because that's what the machine is for. It allows it to be tested with mock services.

Reply:
That's fair, I agree.

One thing that's unclear is what happens when `done` gets triggered more than once? Wouldn't that happen in this case, when the condition holds and both `done` and `creatingPagesStatefully` are activated then `creatingPagesStatefully` will also activate `done`? That's why I asked whether this step wouldn't have the opposite condition as its sibling.

Reply:
No, because the transitions are evaluated in order until one matches, and only that one is executed. In this case, on first run it will transition to `creatingPagesStatefully`, which in turn will transition to `done`, but on subsequent runs it will transition straight to `done`.

Reply:
Ah ok. As point of feedback on that API; that was (clearly) not obvious to me when I read the config object. Maybe naming could improve to make that more intuitive?